lib/agentx.c

```diff
@@ -55,13 +55,29 @@ static int agentx_timeout(struct thread *t)
 static int agentx_read(struct thread *t)
 {
 	fd_set fds;
+	int flags;
+	int nonblock = false;
 	struct listnode *ln = THREAD_ARG(t);
 	list_delete_node(events, ln);
 
+	/* fix for non blocking socket */
+	flags = fcntl(THREAD_FD(t), F_GETFL, 0);
+	if (-1 == flags)
+		return -1;
+
+	if (flags & O_NONBLOCK)
+		nonblock = true;
+	else
+		fcntl(THREAD_FD(t), F_SETFL, flags | O_NONBLOCK);
+
 	FD_ZERO(&fds);
 	FD_SET(THREAD_FD(t), &fds);
 	snmp_read(&fds);
 
+	/* Reset the flag */
+	if (!nonblock)
+		fcntl(THREAD_FD(t), F_SETFL, flags);
```

Comment:
Why are we turning it off again? Is this needed?

Reply:
> Why are we turning it off again? Is this needed?
Please note that the defect was because vtysh was hanging for a blocking snmp socket in bgp. What we are doing here is as follows:
1. when we read from the agent, check the type of the socket. if the socket is blocking, convert it into non blocking socket by setting a non-blocking flag on it. If the socket is already non-blocking, do not do anything.
2. read the data from the socket.
3. while returning from the agent, if we had set the non blocking flag on the socket, we reset that flag. This makes sure that when we return from this function, we have the same state on the socket flag as we had when we entered inside the function.
Hope I was clear.

Reply:
Yes, I understood the code. I was asking why 3) is needed.

Reply:
netsnmp is library code(can change from version to version). It might have some code that depends on the socket flag being blocking or non-blocking. When I changed the code here, I did not want to change the socket flag.

Reply:
Oh, I see. The sockets themselves are created by net-snmp. We get them, make them nonblocking, then pass them to another net-snmp library function. Then set the flag back to whatever it was before. Real gnarly.

Someone with a better understanding of net-snmp should probably take a look at this one.

Reply:
Please have a look at this.
https://sourceforge.net/p/net-snmp/patches/1348/